### ISSUE TITLE
Save DF with nested records with the same name

### ISSUE THREAD

**Issue by sixers** [databricks/spark-avro#54]
spark-avro fails to save this dataset:

```python
df = sqlc.createDataFrame([Row(id=1, data=Row(data=Row(key='value')))])
df.save("s3://...", "com.databricks.spark.avro")
```
with the following error:

```
Py4JJavaError: An error occurred while calling o1873.save.
: org.apache.avro.SchemaParseException: Can't redefine: data
	at org.apache.avro.SchemaBuilder$NameContext.put(SchemaBuilder.java:936)
	at org.apache.avro.SchemaBuilder$NameContext.access$600(SchemaBuilder.java:884)
	at org.apache.avro.SchemaBuilder$NamespacedBuilder.completeSchema(SchemaBuilder.java:470)
	at org.apache.avro.SchemaBuilder$RecordBuilder.fields(SchemaBuilder.java:1734)
	at com.databricks.spark.avro.SchemaConverters$.convertStructToAvro(SchemaConverters.scala:99)
	at com.databricks.spark.avro.SchemaConverters$.com$databricks$spark$avro$SchemaConverters$$convertFieldTypeToAvro(SchemaConverters.scala:190)
	at com.databricks.spark.avro.SchemaConverters$$anonfun$convertStructToAvro$1.apply(SchemaConverters.scala:104)
	at com.databricks.spark.avro.SchemaConverters$$anonfun$convertStructToAvro$1.apply(SchemaConverters.scala:100)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
	at com.databricks.spark.avro.SchemaConverters$.convertStructToAvro(SchemaConverters.scala:100)
	at com.databricks.spark.avro.SchemaConverters$.com$databricks$spark$avro$SchemaConverters$$convertFieldTypeToAvro(SchemaConverters.scala:190)
	at com.databricks.spark.avro.SchemaConverters$$anonfun$convertStructToAvro$1.apply(SchemaConverters.scala:104)
	at com.databricks.spark.avro.SchemaConverters$$anonfun$convertStructToAvro$1.apply(SchemaConverters.scala:100)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
	at com.databricks.spark.avro.SchemaConverters$.convertStructToAvro(SchemaConverters.scala:100)
	at com.databricks.spark.avro.AvroSaver$.save(AvroSaver.scala:66)
	at com.databricks.spark.avro.package$AvroDataFrame.saveAsAvroFile(package.scala:37)
	at com.databricks.spark.avro.DefaultSource.createRelation(DefaultSource.scala:84)
	at org.apache.spark.sql.sources.ResolvedDataSource$.apply(ddl.scala:240)
	at org.apache.spark.sql.DataFrame.save(DataFrame.scala:1196)
	at org.apache.spark.sql.DataFrame.save(DataFrame.scala:1181)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:379)
	at py4j.Gateway.invoke(Gateway.java:259)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:207)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jendap** [comment]
I ran to it as well. So I have spend some time investigating it....

Short answer:
Avro does not allow two records with the same name within schema. So we could rename them but then we would break "union" type resolution. Fixing it will need some work in upstream avro. And a bit of work here.


tl;dr

Long answer:
We could rename the records to use different "name". But then the union resolution would not work! The union resolution is looking up field name from data in record names in schema.

Think of the following data:
```json
{"data":{"data":"value"}}
```

The schema for it would be:
```json
{"type":"record", "name":"data", "fields":[
    {"name":"data","type":[
        {"type":"record","name":"data","fields":[{"name":"data","type":["string","null"]}]},
        "null"
    ]}
]}
```
Note:
 * two "record" with the same "name"
 * field "type" is an array of a record and null - union of type and null is a way avro represent nullable fields

We could rename the inner record name to "data2" and avoid 'Can't redefine: data' exception...
```json
{"type":"record", "name":"data", "fields":[
    {"name":"data","type":[
        {"type":"record","name":"data2","fields":[{"name":"data","type":["string","null"]}]},
        "null"
    ]}
]}
```

But then the data `{"data":{"data":"value"}}` will not work. The outer filed name "data" ( `{"THIS_DATA":{"data":"value"}}`) would be matched against top level record's field name "data" in the schema. That is ok.

The inner field name "data" (`{"data":{"THIS_DATA":"value"}}`) will be looked up in the union. Union of "null" and "data2" (inner record name). The problem is that there is no "data" in ["data2", "null"].

Hence there is really no way to do this currently. The only way I see could be supported with current avro implementation would be without unions (not nullable fields) and we would have to rename records in schema here in spark-avro module. We could generate:
```json
{"type":"record", "name":"data", "fields":[
    {"name":"data","type":
        {"type":"record","name":"data2","fields":[{"name":"data","type":"string"}]}
    }
]}
```
That would work. Note the inner field would be generated with name "data2" instead of "data". Also note that the "type" inside fields is not an array any more as well as there is no "null". Hence no union and no problem with union resolution.

The problem is in upstream implementation. You can't have two records with the same name in one schema. But then the record name is used for union resolution  based on field name in data.

I'm thinking of proposing a solution to upstream using aliases. Think of following...
```json
{"type":"record", "name":"data", "fields":[
    {"name":"data","type":[
        {"type":"record","name":"data2","aliases":["data"],"fields":[{"name":"data","type":["string","null"]}]},
        "null"
    ]}
]}
```
If the union resolution would match data field "data" against record name "data2" OR its alias "data" it should just work.

I'm not sure if this solution will be good enough for upstream. Also avro does not seem to have frequent releases lately. Do not expect a good solution any time soon :(

**Comment by reyven** [comment]
I implemented an alternative solution to this problem using namespaces instead of aliases.

Added a pull request for review:
https://github.com/databricks/spark-avro/pull/73

Basically it involves two steps:
  - concatenate enclosing record names into the namespace of a given record to make its fullname unique
  - declare the name of the record as a fullname as opposed to separate name and namespace declarations (which is allowed according to the Avro specification), so that union type resolutions still work

My only concern is whether this is a proper use of namespaces, and whether it interferes with any of their original purposes.

**Comment by zhangxuhong** [comment]
will this issue be solved, and do have other choice to save to avro format?

**Comment by gvdm90** [comment]
Hi,

I'm having this issue and I have no choice to change neither the fields names nor their class (which is exactly the same, an Enum). 

My JSON structure is this

```
{
	"name": "StatisticsObj",
	"type": "record",
	"fields": [{
			"default": null,
			"name": "requestStatistics",
			"type": ["null", {
					"name": "RequestStatistics",
					"type": "record",
					"fields": [{
							"default": null,
							"name": "operationType",
							"type": ["null", {
									"name": "OperationTypeEnum",
									"type": "enum",
									"symbols": ["JOB", "REQUEST"]
								}
							]
						}
					]
				}
			]
		}, {
			"default": null,
			"name": "jobStatistics",
			"type": ["null", {
					"name": "JobStatistics",
					"type": "record",
					"fields": [{
							"default": null,
							"name": "operationType",
							"type": ["null", {
									"name": "OperationTypeEnum",
									"type": "enum",
									"symbols": ["JOB", "REQUEST"]
								}
							]
						}
					]
				}
			]
		}
	]
}
```

Is there anyway to solve or workaround this issue?

**Comment by gvdm90** [comment]
Ok I solved using namespaces in the Avro record schema
